### ISSUE TITLE
Improve benchmark performance

### ISSUE THREAD

**Issue by cBournhonesque** [cBournhonesque/lightyear#421]
Benchmarks show that it takes 1.3 ms to replicate 1000 entities (replicon takes 30us). Why?

With a lot of spans, it's 3ms (because of the tracing overhead):

- send_entity_spawn takes 530us
  - ReplicationSend::prepare_entity_spawn is 85us (because we are allocating. The overall memory should be re-used though. But the hashmap inside the replication_group_id is probably very inefficient?)

- send_component_update is 686us
  - ReplicationSend::prepare_component_insert is 178us
  - rest is probably iteration + serialization?
  
- networking::send is 1.13ms
  - buffer_replication_message is 790us
    - finalize is 115us
    - then there is serializing (not tracked)
    - buffer_send_with_priority is 170us (buffering into the message manager)
  - send_packets is only 335us
    - message_manager.send_packets (that collects the messages to send from channels and builds the packets) is 190us

**Comment by Nul-led** [comment]
Did you actually run their benchmarks or did you just trust their results? :p

**Comment by cBournhonesque** [comment]
Ran their benchmarks :)
I'm trying to understand why the difference can be so big. Probably because of extra allocations?

**Comment by cBournhonesque** [comment]
First optimization I will try:
- Instead of building a EntityActionsMessage (which allocates to store the component insertion data, and then gets re-serialized) into a buffer, we will try to write the `EntityActionsMessage` directly
  - 2 options: 
    1) have one intermediary buffer for each entity in the replication group, and then we will stitch all the buffers in the end. Requires extra allocations but reduces bandwidth because we don't write the same Entity multiple times.
Basically for each entity, we will store [Spawn, Despawn, Removals, Insertions]
     2) for the replication group, write all [Spawns], then all [Despawns], then all [Removals], then all [Insertions]. This has better performance because we can write the final message directly in one go without intermediary buffers, but it requires more bandwidth because we might have to write the `Entity` multiple times (once in Spawn, once in Removals, etc.)
     
Note that these new approaches also probably require more bandwidth, because `bitcode` was able to use bit-compression for the previous `EntityActionsMessage`.

**Comment by Nul-led** [comment]
imo bandwidth is still the most important thing to optimize for multiplayer games. If its impossible to optimize the serialization due to an increase in bandwidth, having this tradeoff seems worth it. Ofc finding a balance is a good idea but i feel like optimizing for bandwidth first makes more sense.

**Comment by cBournhonesque** [comment]
Agreed that bandwidth is more important but this is still a massive difference: 40X !
I would accept something like 5 or 10X (i.e. 300us to serialize the 1000 entities), but 1.3ms is way too much


I think it's mostly due to 2 things:
- they send the EntityActions in one big packet, so they can serialize it directly. It works but then the packet is split up into multiple parts, which adds a bunch of latency if any of the fragment is dropped or delayed. Instead we send a lot of individual messages, which is more flexible but we use more CPU because we keep some data around in memory. (but maybe we can pre-allocate some memory or something?)
- they serialize their message directly while because of bitcode I serialize stuff twice, and have extra allocations. In general I think I want to move away from bitcode. It's good for bandwidth but it's too much of a headache compared with an easier serialization approach

**Comment by Nul-led** [comment]
Do you know how well lightyear compare to them based on avg bandwidth usage? Might be interesting to have benchmarks for that too..

Apart from that
1) sounds like a massive pain to resolve and would probably require huge internal changes as far as i can tell
2) what is your definition of "easier" serialization?

**Comment by cBournhonesque** [comment]
1. Is actually much easier to do than what i do now. Right now I do some complicated stuff to make sure that EntityActions are split up instead of sending them all as a giant message. I think the first thing i'll try is actually to just send all EntityActions in one giant message where I serialize stuff directly inside.
2. Something that uses the `Read`/`Write` traits, and doesn't use bit-level stuff

**Comment by cBournhonesque** [comment]
As a simple first step, replicating everything as a single `ReplicationGroup` brings the time to around 780us, which is a huge improvement. Probably because we don't allocate new space in the hashmaps, and we the vec allocations are more efficient.

Also instead instead of sending lots of small messages we send one big message
